transport/peer/peerlist/peerring.go

```diff
@@ -0,0 +1,181 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package peerlist
+
+import (
+	"container/ring"
+	"sync"
+
+	"go.uber.org/yarpc/transport"
+	"go.uber.org/yarpc/transport/internal/errors"
+)
+
+// NewPeerRing creates a new PeerRing with initial size of "length"
+func NewPeerRing(length int) *PeerRing {
+	return &PeerRing{
+		peerToNode: make(map[string]*peerRingNode, length),
+	}
+}
+
+// PeerRing provides a safe way to interact (Add/Remove/Get) with a potentially
+// changing list of peer objects
+type PeerRing struct {
+	lock sync.Mutex
+
+	peerToNode map[string]*peerRingNode
+	nextNode   *peerRingNode
+}
+
+// Add a transport.Peer to the end of the PeerRing, if the ring is empty
+// it initializes the nextNode marker
+func (pr *PeerRing) Add(peer transport.Peer) error {
+	pr.lock.Lock()
+	defer pr.lock.Unlock()
+
+	if _, ok := pr.peerToNode[peer.Identifier()]; ok {
+		// Peer Already in ring, ignore the add
+		return errors.ErrPeerAlreadyInList{
+			Peer: peer,
+		}
+	}
+
+	newNode := newPeerRingNode(peer)
+	pr.peerToNode[peer.Identifier()] = newNode
+
+	if pr.nextNode == nil {
+		// Empty ring, add the first node
+		pr.nextNode = newNode
+	} else {
+		// Push the node to the ring
+		pr.nextNode.push(newNode)
```

Comment:
consider inlining push. the indirection makes it less clear that this does in fact add the node before the cursor, not after